### PR TITLE
Call math functions from UDFs

### PR DESCRIPTION
**Pull request by guilhermeleobas**
As the title suggests, this PR makes math functions (and possibly others) available to UDFs.

Still a work in progress.

- [x] Build a list of functions that numba/OmniSci supports
- [x] Investigate the cause of segfaults. See 3 comments below.
- [ ] Fix segfaults
- [ ] Implement code for returning an array from a function.
- [ ] remove `@llvm.lifetime.start.p0i8` and `@llvm.lifetime.end.p0i8` from the IR.

### CODE REVIEW

**Comment by guilhermeleobas** [comment]
I will use this post to update the functions that works:

### Trigonometric functions
- [x] `np.sin(x)`
- [x] `np.cos(x)`
- [x] `np.tan(x)`
- [x] `np.arcsin(x)`
- [x] `np.arccos(x)`
- [x] `np.arctan(x)`
- [ ] `np.hypot(x,y)` # fails because it needs codegen support
- [x] `np.arctan2(x,y)`

### Hyperbolic functions
- [ ] `np.sinh(x)` # omnisci segfaults
- [ ] `np.cosh(x)` # omnisci segfaults
- [ ] `np.tanh(x)` # segfaults - numba impl uses sinh/cosh
- [x] `np.arcsinh(x)`
- [x] `np.arccosh(x)`
- [x] `np.arctanh(x)`

### Rounding
- [ ] `np.around(x)`
- [ ] `np.round_(x)`
- [ ] `np.rint(x)`
- [ ] `np.fix(x)`
- [ ] `np.floor(x)`
- [ ] `np.ceil(x)`
- [ ] `np.trunc(x)`

### Sums, products, differences
- [ ] `np.prod(x)`
- [ ] `np.sum(x)`
- [ ] `np.nanprod(x)`
- [ ] `np.nansum(x)`
- [ ] `np.cumprod(x)`
- [ ] `np.cumsum(x)`
- [ ] `np.nancumprod(x)`
- [ ] `np.nancumsum(x)`
- [ ] `np.diff(x)`
- [ ] `np.ediff1d(x)`
- [ ] `np.gradient(x)`
- [ ] `np.cross(x)`
- [ ] `np.trapz(x)`

### Expoents and logarithms
- [ ] `np.exp(x)`
- [ ] `np.expm1(x)`
- [ ] `np.exp2(x)`
- [ ] `np.log(x)`
- [ ] `np.log10(x)`
- [ ] `np.log2(x)`
- [ ] `np.log1p(x)`
- [ ] `np.logaddexp(x)`
- [ ] `np.logaddexp2(x)`

**Comment by guilhermeleobas** [comment]
Hey @pearu,

I spend these last couple of days trying to compile the methods that need codegen support (`np.hypot`, `np.lcm` and `np.gcd`). 

Take for instance the `hypot_float_impl` method:

```python
@lower(math.hypot, types.Float, types.Float)
def hypot_float_impl(context, builder, sig, args):
    xty, yty = sig.args
    assert xty == yty == sig.return_type
    x, y = args

    # Windows has alternate names for hypot/hypotf, see
    # https://msdn.microsoft.com/fr-fr/library/a9yb3dbt%28v=vs.80%29.aspx
    fname = {
        types.float32: "_hypotf" if sys.platform == 'win32' else "hypotf",
        types.float64: "_hypot" if sys.platform == 'win32' else "hypot",
    }[xty]
    plat_hypot = types.ExternalFunction(fname, sig)

    if sys.platform == 'win32' and config.MACHINE_BITS == 32:
        inf = xty(float('inf'))

        def hypot_impl(x, y):
            if math.isinf(x) or math.isinf(y):
                return inf
            return plat_hypot(x, y)
    else:
        def hypot_impl(x, y):
            return plat_hypot(x, y)

    res = context.compile_internal(builder, hypot_impl, sig, args)
    return impl_ret_untracked(context, builder, sig.return_type, res)
```

`hypot_float_impl` calls `context.compile_internal` which calls `RuntimeLinker` methods. Since rbc [replaces the `RuntimeLinker` with `None` ](https://github.com/xnd-project/rbc/blob/f2b12bf6ca62a31ac3e146602f86d82ef71bf318/rbc/irtools.py#L40), numba ends up calling a method of a `None` object. 

I think we can resolve this particular issue by creating a dummy linker with the same methods `RuntimeLinker` but empty. What do you think?

```python
class RemoteRuntimeLinker(nb.targets.codegen.RuntimeLinker):

    def scan_unresolved_symbols(self, module, engine):
        pass

    def scan_defined_symbols(self, module):
        pass

    def resolve(self, engine):
        pass

class JITRemoteCPUCodegen(nb.targets.codegen.JITCPUCodegen):
    # TODO: introduce JITRemoteCodeLibrary?
    _library_class = nb.targets.codegen.JITCodeLibrary

    def __init__(self, name, target_info):
        self.target_info = target_info
        super(JITRemoteCPUCodegen, self).__init__(name)
        # we cannot use local host RuntimeLinker, in general.
        self._rtlinker = RemoteRuntimeLinker()
```

**Comment by guilhermeleobas** [comment]
```
; ModuleID = 'rbc.irtools.compile_to_IR'
source_filename = "<string>"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v32:32:32-v64:64:64-v128:128:128-n16:32:64"
target triple = "nvptx64-nvidia-cuda"
@"_ZN08NumbaEnv8__main__8sinh$244Ed" = common local_unnamed_addr global i8* null
; Function Attrs: nounwind writeonly
define i32 @"_ZN8__main__8sinh$244Ed"(double* noalias nocapture %retptr, { i8*, i32 }** noalias nocapture readnone %excinfo, double %arg.x) local_unnamed_addr #0 {
entry:
  %.3.i = alloca double, align 8
  %0 = bitcast double* %.3.i to i8*
  call void @llvm.lifetime.start.p0i8(i64 8, i8* nonnull %0)
  store double 0.000000e+00, double* %.3.i, align 8
  %.7.i = call i32 @"_ZN8__main__8sinh$244Ed"(double* nonnull %.3.i, { i8*, i32 }** undef, double %arg.x) #2
  %1 = bitcast double* %.3.i to i64*
  %.17.i1 = load i64, i64* %1, align 8
  call void @llvm.lifetime.end.p0i8(i64 8, i8* nonnull %0)
  %2 = bitcast double* %retptr to i64*
  store i64 %.17.i1, i64* %2, align 8
  ret i32 0
}
; Function Attrs: nounwind writeonly
define double @sinh(double %.1) local_unnamed_addr #0 {
entry:
  %.3 = alloca double, align 8
  store double 0.000000e+00, double* %.3, align 8
  %.7 = call i32 @"_ZN8__main__8sinh$244Ed"(double* nonnull %.3, { i8*, i32 }** undef, double %.1)
  %.17 = load double, double* %.3, align 8
  ret double %.17
}
; Function Attrs: argmemonly nounwind
declare void @llvm.lifetime.start.p0i8(i64, i8* nocapture) #1
; Function Attrs: argmemonly nounwind
declare void @llvm.lifetime.end.p0i8(i64, i8* nocapture) #1
attributes #0 = { nounwind writeonly }
attributes #1 = { argmemonly nounwind }
attributes #2 = { nounwind }
```

We shouldn't send `@llvm.lifetime.start.p0i8` and `@llvm.lifetime.end.p0i8` to the server.

**Comment by pearu** [comment]
Also, note that `_ZN8__main__8sinh$244Ed` is calling itself.

**Comment by guilhermeleobas** [comment]
I think the problem of all those segmentation faults is the infinite recursion. I ran the server with gdb and the following query inside omnisql: `select x, trunc(x) from table;`

Below is the stacktrace from gdb:
```
[New Thread 0x7fffcb7fe700 (LWP 35285)]

Thread 45 "omnisci_server" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffcb7fe700 (LWP 35285)]
0x00007fffe85991eb in trunc ()
(gdb) where
#0  0x00007fffe85991eb in trunc ()
#1  0x00007fffe85991e0 in ?? ()
#2  0x00007fffe85991ed in trunc ()
#3  0x00007fffe85991e0 in ?? ()
#4  0x00007fffe85991ed in trunc ()
#5  0x00007fffe85991e0 in ?? ()
#6  0x00007fffe85991ed in trunc ()
#7  0x00007fffe85991e0 in ?? ()
#8  0x00007fffe85991ed in trunc ()
#9  0x00007fffe85991e0 in ?? ()
#10 0x00007fffe85991ed in trunc ()
#11 0x00007fffe85991e0 in ?? ()
#12 0x00007fffe85991ed in trunc ()
#13 0x00007fffe85991e0 in ?? ()
#14 0x00007fffe85991ed in trunc ()
#15 0x00007fffe85991e0 in ?? ()
#16 0x00007fffe85991ed in trunc ()
#17 0x00007fffe85991e0 in ?? ()
#18 0x00007fffe85991ed in trunc ()
#19 0x00007fffe85991e0 in ?? ()
#20 0x00007fffe85991ed in trunc ()
#21 0x00007fffe85991e0 in ?? ()
#22 0x00007fffe85991ed in trunc ()
#23 0x00007fffe85991e0 in ?? ()
#24 0x00007fffe85991ed in trunc ()
#25 0x00007fffe85991e0 in ?? ()
#26 0x00007fffe85991ed in trunc ()
#27 0x00007fffe85991e0 in ?? ()
#28 0x00007fffe85991ed in trunc ()
#29 0x00007fffe85991e0 in ?? ()
#30 0x00007fffe85991ed in trunc ()
#31 0x00007fffe85991e0 in ?? ()
#32 0x00007fffe85991ed in trunc ()
#33 0x00007fffe85991e0 in ?? ()
#34 0x00007fffe85991ed in trunc ()
#35 0x00007fffe85991e0 in ?? ()
#36 0x00007fffe85991ed in trunc ()
#37 0x00007fffe85991e0 in ?? ()
#38 0x00007fffe85991ed in trunc ()
#39 0x00007fffe85991e0 in ?? ()
#40 0x00007fffe85991ed in trunc ()
#41 0x00007fffe85991e0 in ?? ()
#42 0x00007fffe85991ed in trunc ()
#43 0x00007fffe85991e0 in ?? ()
#44 0x00007fffe85991ed in trunc ()
...
```

So yeah, `trunc` keeps calling itself.

**Comment by guilhermeleobas** [comment]
Added a list of[ _forbidden function_ ](https://github.com/xnd-project/rbc/issues/32)names and rbc now launches an error if the user tries to define a function with such name. I've also changed how rbc behaves when one tries to define the same function twice using omnisci as the backend.

**Comment by guilhermeleobas** [comment]
@pearu, when you have time, can you please review this PR? 

Obs: The trick of adding a prefix on function names (`_trunc`, `_sinh`) doesn't work for some functions: 

```python
def _exp2(x):
    return np.exp2(x)
```

File: rbc/tests/test_omnisci_math.py [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+    omnisci.register()
+
+    for fn_name in ['around', '_rint', 'round_', 'floor', 'ceil', '_trunc']:
+        np_fn = getattr(np, fn_name.lstrip('_'))
+
+        query = ''
+        if fn_name == 'arccosh':
+            query = 'select i, {fn_name}(i) from {omnisci.table_name}'\
+                    .format(**locals())
+        else:
```
Comment (pearu) [inline rbc/tests/test_omnisci_math.py]: Delete this block.

File: rbc/tests/test_omnisci_math.py [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+
+    @omnisci('double(double)')  # noqa: F811
+    def exp(x):
+        return np.exp(x)
+
+    @omnisci('double(double)')  # noqa: F811
+    def _expm1(x):
+        return np.expm1(x)
+
+    # doesn't work - even adding a prefix '_'
```
Comment (pearu) [inline rbc/tests/test_omnisci_math.py]: Does it work when using other prefixes? Say, `_np`?
Open a separate rbc issue for this.

File: rbc/tests/test_omnisci_math.py [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+
+    @omnisci('double(double)')  # noqa: F811
+    def log(x):
+        return np.log(x)
+
+    @omnisci('double(double)')  # noqa: F811
+    def log10(x):
+        return np.log10(x)
+
+    # doesn't work - even adding a prefix '_'
```
Comment (pearu) [inline rbc/tests/test_omnisci_math.py]: Add to the exp2 issue.

File: rbc/tests/test_omnisci_math.py [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+    # @omnisci('double(double)')  # noqa: F811
+    # def _log2(x):
+    #     return np.log2(x)
+
+    @omnisci('double(double)')  # noqa: F811
+    def _log1p(x):
+        return np.log1p(x)
+
+    # doesn't work - even adding a prefix '_'
```
Comment (pearu) [inline rbc/tests/test_omnisci_math.py]: Add to the exp2 issue.

File: rbc/tests/test_omnisci_math.py [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+    # @omnisci('double(double, double)')  # noqa: F811
+    # def _logaddexp(x, y):
+    #     return np.logaddexp(x, y)
+
+    # doesn't work - even adding a prefix '_'
```
Comment (pearu) [inline rbc/tests/test_omnisci_math.py]: Add to the exp2 issue.

File: rbc/tests/test_omnisci_math.py [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+
+    for fn_name in ['_sum', '_prod']:
+        np_fn = getattr(np, fn_name.lstrip('_'))
+
+        descr, result = omnisci.sql_execute(
+            'select t, {fn_name}(t) from {omnisci.table_name}'
+            .format(**locals()))
+
+        for t, v in list(result):
+            assert(np.isclose(np_fn(t), v))
```
Comment (pearu) [inline rbc/tests/test_omnisci_math.py]: Use `assert(np_fn(t) == v)` as `t` is integer array.

File: rbc/tests/test_omnisci_math.py [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+    def floor_divide(x, y):
+        return np.floor_divide(x, y)
+
+    @omnisci('double(double, double)')
+    def _fmod(x, y):
+        return np.fmod(x, y)
+
+    @omnisci('int(int, int)')  # noqa: F811
+    def mod(x, y):
+        return np.mod(x, y)
```
Comment (pearu) [inline rbc/tests/test_omnisci_math.py]: `np.mod` works also for double inputs. So, apply also `@omnisci('double(double, double)')` to `mod`?

File: rbc/tests/test_omnisci_math.py [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+        d = np.power(c, c)
+        return np.trunc(d)
+
+    omnisci.register()
+
+    _, result = omnisci.sql_execute(
+        'select x, x, multiple1(x, y) from {omnisci.table_name}'
+        .format(**locals()))
+
+    expected = [0.0, 0.0, 0.0, 1.0, 2.0]
```
Comment (pearu) [inline rbc/tests/test_omnisci_math.py]: Use
```
expected = multiple1.func(np.arange(1, 6)/10.0)
```

File: rbc/tests/test_remotejit.py [hunk 9]
```diff
@@ -171,6 +172,23 @@ def add(a, b):
...
 
 
+def test_rjit_np(rjit):
+
+    @rjit('double(double)')
+    def trunc(x):
+        return np.trunc(x)
+
+    assert trunc(0.3) == 0.0
+    assert trunc(2.9) == 2.0
```
Comment (pearu) [inline rbc/tests/test_remotejit.py]: Use `assert(np.allclose(trunc(0.3), np.trunc(0.3)))` etc.

File: rbc/tests/test_remotejit.py [hunk 10]
```diff
@@ -171,6 +172,23 @@ def add(a, b):
...
+
+    @rjit('double(double)')
+    def exp2(x):
+        return np.exp2(x)
+
+    assert exp2(2) == 4.0
+    assert exp2(3) == 8.0
```
Comment (pearu) [inline rbc/tests/test_remotejit.py]: Use assert(np.allclose(exp2(2), np.exp2(2))) etc.

File: rbc/tests/test_omnisci_math.py [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+
+    omnisci.register()
+
+    for fn_name in ['_sinh', '_cosh', '_tanh', 'arcsinh',
+                    'arccosh', 'arctanh']:
+        np_fn = getattr(np, fn_name.lstrip('_'))
+
+        query = ''
+        if fn_name == 'arccosh':
+            query = 'select i, {fn_name}(i) from {omnisci.table_name}'\
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply rbc/tests/test_omnisci_math.py]: Changed the code to test with float values as well

**Comment by guilhermeleobas** [comment]
Test `test_omnisci_array.py::test_even_sum` is failing on my machine with the following error:

```
Exception: Function array_even_sum_int32(BOOLEAN[], INTEGER[]) not supported.
 Existing extension function implementations:
    array_even_sum_int32({i8*, i64, i8}*, {i32*, i64, i8}*) -> i32
```

**Comment by pearu** [comment]
You are likely using the server from omniscidb-internal master which thrift configuration is different from the one in rbc. To pass the test, you must use omniscdb version < 5.2.

Fixing https://github.com/xnd-project/rbc/issues/18 should resolve this issue.

**Comment by guilhermeleobas** [comment]
Last commit fixes #36

**Comment by guilhermeleobas** [comment]
@pearu, I think this is ready for a final review. I had to include some more stuff to uncomment lines 170-171 https://github.com/xnd-project/rbc/blob/5a55c5aff6327bfb98988acc0ff3d1be4645a358/rbc/irtools.py#L170-L171

**Pull request merged**